Reduce of `free()` in the project

cdumez opened the pull request:
#### 59ab652174061d307b08c1e631a7883560f09615
<pre>
Reduce of `free()` in the project
<a href="https://bugs.webkit.org/show_bug.cgi?id=300491">https://bugs.webkit.org/show_bug.cgi?id=300491</a>

Reviewed by Ryosuke Niwa and Darin Adler.

Reduce of `free()` in the project as manual memory management is dangerous.

* Source/JavaScriptCore/API/JSValue.mm:
(createStructHandlerMap):
* Source/JavaScriptCore/API/JSWrapperMap.mm:
(parsePropertyAttributes):
* Source/JavaScriptCore/disassembler/Disassembler.cpp:
* Source/JavaScriptCore/jsc.cpp:
(runInteractive):
* Source/WTF/wtf/ObjCRuntimeExtras.h:
* Source/WTF/wtf/ObjCRuntimeExtras.mm:
(WTF::property_copyAttributeListSpan):
* Source/WTF/wtf/StackTrace.h:
* Source/WTF/wtf/SystemFree.h:
(WTF::SystemFree::operator() const):
(WTF::adoptSystem):
(WTF::SystemFree&lt;T::operator() const): Deleted.
* Source/WebKit/NetworkProcess/webrtc/NetworkRTCUDPSocketCocoa.mm:
(WebKit::socketAddressFromIncomingConnection):
* Source/WebKit/Platform/cocoa/NetworkIssueReporter.h:
* Source/WebKit/Platform/cocoa/NetworkIssueReporter.mm:
(WebKit::NetworkIssueReporter::NetworkIssueReporter):
(WebKit::NetworkIssueReporter::report):
(WebKit::NetworkIssueReporter::~NetworkIssueReporter): Deleted.
* Source/WebKit/Shared/Cocoa/LaunchLogHook.mm:
(WebKit::LaunchLogHook::initialize):
* Source/WebKit/WebProcess/cocoa/WebProcessCocoa.mm:
(WebKit::registerLogClient):
* Source/WebKitLegacy/mac/WebView/WebPDFView.mm:
(-[WebPDFView _temporaryPDFDirectoryPath]):

Canonical link: <a href="https://commits.webkit.org/301351@main">https://commits.webkit.org/301351@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/60f38cce9c755231ac6d2d9be615a725721c7101

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/125676 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/45338 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/36088 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/132540 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/77559 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/65f0fbb1-31a8-4344-b9f3-0617c1d8d1a5) 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/46023 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/53897 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/95749 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/63864 "Passed tests") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/2e99c9e1-74a3-4a77-8093-57848027209e) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/128624 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/36795 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/112385 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/76241 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/d80a244f-dcc2-4c22-a386-3acbd32ecd13) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/35697 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/30568 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/76009 "Built successfully") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/117761 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/106573 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/30784 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/135214 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/124185 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/52468 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/40230 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/104214 "3 flakes 57 failures") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/52915 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/108596 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/103943 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/26475 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/49304 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/27611 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/49707 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/52363 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/58168 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/157201 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/51710 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/39340 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/55062 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/53407 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WTF/wtf/SystemFree.h:
```diff
@@ -40,19 +40,11 @@ struct SystemFree {
 
 } // namespace WTF
+
+using WTF::adoptSystem;
```
By the way, I'm not a huge fan of this name. I don't find it descriptive enough. If someone has a better proposal, please let me know.

rniwa replied:
Why not change the template to look more like `makeUnique`?
Then we can call this function `makeUniqueWithSystemMalloc` or something.

cdumez replied:
makeUnique would make it look like it allocates memory, whereas this adopts it.

geoffreygaren replied:
I also find ‘System’ to be too broad. Maybe ‘adoptSystemMalloc’? We can also do ‘adoptFastMalloc’ if needed.

cdumez replied:
Yes, I like `adoptSystemMalloc` better. I'll upload a PR to do the rename, thanks!

cdumez replied:
https://github.com/WebKit/WebKit/pull/52205

darinadler commented on Source/JavaScriptCore/jsc.cpp:
```diff
@@ -3917,18 +3917,15 @@ static void runInteractive(GlobalObject* globalObject)
             if (!line)
                 break;
-            source = makeString(source, String::fromUTF8(line), '\n');
+            source = makeString(source, String::fromUTF8(line.get()), '\n');
```
Something I noticed in passing: If we cared about performance here, which we might not, we’d use a `std::span<const char8_t>` instead of creating and destroying a temporary `String`. Might be as simple as `byteCast<char8_t>(unsafeSpan(line.get()))`.

darinadler commented on Source/WTF/wtf/SystemFree.h:
```diff
@@ -31,7 +31,7 @@ namespace WTF {
 template<typename T>
 struct SystemFree {
-    static_assert(std::is_trivially_destructible<T>::value);
+    static_assert(std::is_trivially_destructible<T>::value || std::is_void_v<T>);
```
Since we’re touching this line, I would use `_v` for the old part too.
```suggestion
    static_assert(std::is_trivially_destructible_v<T> || std::is_void_v<T>);
```

cdumez committed:
Reduce of `free()` in the project
https://bugs.webkit.org/show_bug.cgi?id=300491

Reviewed by Ryosuke Niwa and Darin Adler.

Reduce of `free()` in the project as manual memory management is dangerous.

* Source/JavaScriptCore/API/JSValue.mm:
(createStructHandlerMap):
* Source/JavaScriptCore/API/JSWrapperMap.mm:
(parsePropertyAttributes):
* Source/JavaScriptCore/disassembler/Disassembler.cpp:
* Source/JavaScriptCore/jsc.cpp:
(runInteractive):
* Source/WTF/wtf/ObjCRuntimeExtras.h:
* Source/WTF/wtf/ObjCRuntimeExtras.mm:
(WTF::property_copyAttributeListSpan):
* Source/WTF/wtf/StackTrace.h:
* Source/WTF/wtf/SystemFree.h:
(WTF::SystemFree::operator() const):
(WTF::adoptSystem):
(WTF::SystemFree<T::operator() const): Deleted.
* Source/WebKit/NetworkProcess/webrtc/NetworkRTCUDPSocketCocoa.mm:
(WebKit::socketAddressFromIncomingConnection):
* Source/WebKit/Platform/cocoa/NetworkIssueReporter.h:
* Source/WebKit/Platform/cocoa/NetworkIssueReporter.mm:
(WebKit::NetworkIssueReporter::NetworkIssueReporter):
(WebKit::NetworkIssueReporter::report):
(WebKit::NetworkIssueReporter::~NetworkIssueReporter): Deleted.
* Source/WebKit/Shared/Cocoa/LaunchLogHook.mm:
(WebKit::LaunchLogHook::initialize):
* Source/WebKit/WebProcess/cocoa/WebProcessCocoa.mm:
(WebKit::registerLogClient):
* Source/WebKitLegacy/mac/WebView/WebPDFView.mm:
(-[WebPDFView _temporaryPDFDirectoryPath]):

Canonical link: https://commits.webkit.org/301351@main
